oclif/core
# feat: support plugins with wildcards

Allow plugins to be defined with [minimatch](https://www.npmjs.com/package/minimatch) patterns. For instance: 

```json
{
  "dependencies": {
    "@oclif/core": "^3",
    "@oclif/plugin-help": "^6",
    "@oclif/plugin-plugins": "^4",
    "@org/cool-plugin": "^1",
    "@org/awesome-plugin": "^1",
   },
  "oclif": {
    "plugins": [
        "@oclif/plugin-*",
        "@org/*",
     ]
  }
}
```

This PR also exposes a new option on `Config` called `pluginAdditions` that informs `Config` of additional plugins that should be loaded. This is solves the use case described in https://github.com/oclif/core/pull/976

`pluginAdditions` is defined as:
```
  pluginAdditions?: {
    core?: string[] // core plugins that should be loaded
    dev?: string[] // dev plugins that should be loaded
    path?: string // path to load these plugins from (defaults to root plugin's path)
  }
```

Example:
```typescript
  const config = new Config({
    pluginAdditions: {
      core: ['@oclif/plugin-*'],
    },
    root: __dirname,
  })
  await config.load()
```

**QA**
- checkout branch and `yarn && yarn build`
- `oclif generate my-cli`
- `cd my-cli && yarn link @oclif/core`
- Change `plugins` entry in package.json to `"plugins": ["@oclif/plugin-*"]`
- `bin/dev.js` should should `help` and `plugins` topics 
- `bin/dev.js plugins --core` should show plugin-help and plugin-plugins
- remove `plugins` from package.json
- change `bin/run.js` to:
```javascript
#!/usr/bin/env node

// eslint-disable-next-line unicorn/prefer-top-level-await
;(async () => {
  const {execute} = await import('@oclif/core')
  await execute({
    loadOptions: {
      pluginAdditions: {
        core: ['@oclif/plugin-*'],
      },
      root: __dirname,
    },
  })
})()
```
- `bin/dev.js` should should `help` and `plugins` topics 
- `bin/dev.js plugins --core` should show plugin-help and plugin-plugins
- `oclif generate another-cli`
- `cd another-cli && yarn add @oclif/plugin-version`
- change `bin/dev.js` of `my-cli` to:
```javascript
#!/usr/bin/env node

// eslint-disable-next-line unicorn/prefer-top-level-await
;(async () => {
  const {execute} = await import('@oclif/core')
  await execute({
    loadOptions: {
      pluginAdditions: {
        core: ['@oclif/plugin-*'],
        path: 'path/to/another-cli'
      },
      root: __dirname,
    },
  })
})()
```
- `bin/dev.js` should should `help`, `plugins`, and `version` topics/commands 
- `bin/dev.js plugins --core` should show plugin-help, plugin-plugins, and plugin-version

Closes #976 
Closes https://github.com/oclif/oclif/discussions/1235
